contrib/firebaseio/src/main/java/utils/FirebaseCheckpoint.java

```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright (c) 2015 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not  use this file except  in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package utils;
+
+import com.google.cloud.dataflow.sdk.io.UnboundedSource.CheckpointMark;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import io.FirebaseSource.FirebaseReader;
+
+import org.joda.time.Instant;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.PriorityBlockingQueue;
+
+/**
+ * Maintains a {@link PriorityBlockingQueue} to store the {@link Record}s created by
+ * {@link FirebaseReader}, on error these records are replayed on the new Reader to prevent
+ * dataloss.
+ * @param <T>
+ */
+public class FirebaseCheckpoint<T> implements CheckpointMark {
+
+  //Oldest on top
+  private PriorityBlockingQueue<Record<T>> unread;
+  private LinkedBlockingQueue<Record<T>> read;
+
+  private Record<T> cur;
+
+  @JsonCreator
+  public FirebaseCheckpoint(
+      @JsonProperty("read") Record<T>[] read,
+      @JsonProperty("unread") Record<T>[] unread,
+      @JsonProperty("cur") Record<T> cur){
+    this.read = new LinkedBlockingQueue<>(read.length);
+    this.read.addAll(Arrays.asList(read));
+    this.unread = new PriorityBlockingQueue<>(unread.length);
+    this.unread.addAll(Arrays.asList(unread));
+    this.cur = cur;
+  }
+
+  public FirebaseCheckpoint(){
+    this.unread = new PriorityBlockingQueue<>();
+    this.read = new LinkedBlockingQueue<>();
+  }
+
+  @Override
+  public void finalizeCheckpoint() throws IOException {
+    this.read = new LinkedBlockingQueue<>();
+  }
+
+  public Instant minTimestamp(){
+    if (cur.compareTo(unread.peek()) < 0){
+      return cur.timestamp;
+    } else {
+      return unread.peek().timestamp;
+    }
```

Comment:
I suspect this sort of evens out with the way that Firebase works. If we recieve a bunch of events at once, it's more likely they will be out of order (since Firebase actually operates on state synchronization rather than in order event streams). However, if we receive one at a time they will likely be in order (i.e. if there is one event per synchronization loop).

If there are problems I'll revisit.